src/packages/version-check/src/index.ts

```diff
@@ -0,0 +1,291 @@
+};
+
+export class VersionCheck {
+  protected ConfigManager;
+  protected _config;
+  protected _logger;
+  protected _currentVersion;
+
+  constructor(
+    currentVersion: string,
+    config?: VersionCheckConfig,
+    logger?: Function
+  ) {
+    // Creates a new config, or reads existing from disk
+    this.ConfigManager = new Conf({
+      configName: process.env.TEST ? "testConfig" : "config", // config is the package default
+      defaults: {
+        ...VersionCheck.DEFAULTS,
+        ...config
+      }
+    });
+
+    // pulls the config out of the manager, lays optional props over top
+    this._config = { ...this.ConfigManager.get("config"), ...config };
+
+    // If config was passed in, save changes
+    if (config) this.saveConfig();
+
+    if (this.isValidSemver(currentVersion)) {
+      this._currentVersion = this.cleanSemver(currentVersion);
+    } else {
+      // Semver is invalid, turn off version check
+      this._currentVersion = false;
+      this._config.enabled = false;
+    }
+
+    this._logger = logger || console;
+  }
+
+  init() {
+    this.getLatestVersion();
+    return this;
+  }
+
+  cleanSemver(semver) {
+    return semverClean(semver);
+  }
+
+  isValidSemver(semver) {
+    return semverValid(semver);
+  }
+
+  setEnabled(enabled) {
+    this.set("enabled", enabled);
+  }
+
+  setLatestVersion(latestVersion) {
+    if (this.isValidSemver(latestVersion)) {
+      this.set("latestVersion", latestVersion);
+    }
+  }
+
+  setLatestVersionLogged(latestVersionLogged) {
+    this.set("latestVersionLogged", latestVersionLogged);
+  }
+
+  setPackageName(packageName) {
+    this.set("packageName", packageName);
+  }
+
+  setTTL(ttl) {
+    this.set("ttl", ttl);
+  }
+
+  setUrl(url) {
+    this.set("url", url);
+  }
+
+  private set(key, value) {
+    this._config[key] = value;
+    this.saveConfig();
+  }
+
+  private saveConfig() {
+    this.ConfigManager.set("config", this._config);
+  }
+
+  configFileLocation() {
+    return this.ConfigManager.path;
+  }
+
+  // Intentionally verbose here if we get logging involved it could aid debugging
+  canNotifyUser() {
+    const currentVersion = this._currentVersion;
+    const latestVersion = this._config.latestVersion;
+    // No currentVersion version passed in
+    if (!currentVersion) {
+      return false;
+    } else if (this.alreadyLoggedThisVersion()) {
+      return false;
+    }
+    // returns falsy if function cannot detect semver difference
+    return !!this.detectSemverChange(currentVersion, latestVersion);
+  }
+
+  detectSemverChange(currentVersion, latestVersion) {
+    if (
+      !currentVersion ||
+      !latestVersion ||
+      !this.isValidSemver(currentVersion) ||
+      !this.isValidSemver(latestVersion) ||
+      semverGte(currentVersion, latestVersion)
+    )
+      return null;
+
+    return semverDiff(currentVersion, latestVersion);
+  }
+
+  async getLatestVersion() {
+    if (!this._config.enabled) return false;
```

Comment:
Should we ping if we are running in CI? :thinking:

Reply:
I don't think we should put any sort of funnel on it for now. 

`ci-detect` just sniffs a bunch of env vars. It wouldn't be hard to maintain our own list but do we want a piece of code that will sniff the env vars, at most, once a day? I am not inclined in favor of that.